File: code/02-static-vs-dynamic-languages/motorcycles.swift
// Run with swiftc motorcycles.swift -o mc
import Foundation

enum MotorcycleType: String {
    case sport
    case naked
    case touring
    case adventure
    case dualSport = "dual_sport"
    case motocross
    case crossCountry
    case trail
}

class Motorcycle {
    let model: String
    let style: MotorcycleType
    let engineSize: Int
    let offRoad: Bool?

    init(model: String, style: MotorcycleType, engineSize: Int, offRoad: Bool) {
        self.model = model
        self.style = style
        self.engineSize = engineSize
        self.offRoad = offRoad
    }

    var canJump: Bool {
        return offRoad!
    }

    static func createAdventure(model: String, engineSize: Int) -> Motorcycle {
        let bike = Motorcycle(model: model, style: .adventure, engineSize: engineSize, offRoad: true)
        return bike
    }

    var description: String {
        return "\(model) \(engineSize), type: \(style.rawValue), off-road: \(offRoad!), can jump: \(canJump)"
    }
}

func createSomeBikes() -> [Motorcycle] {
    let motorcycles = [
        Motorcycle.createAdventure(model: "Himalayan", engineSize: 410),
        Motorcycle.createAdventure(model: "Tenere", engineSize: 700),
        Motorcycle.createAdventure(model: "KTM", engineSize: 790),
        Motorcycle.createAdventure(model: "Norden", engineSize: 901)
    ]

    return motorcycles
}

print("Some bikes:")
let bikes = createSomeBikes()
for b in bikes {
    print(b.description)
}

File: code/02-static-vs-dynamic-languages/motorcycles.cs
using System;
using System.Collections.Generic;

// Run with new project via:
// 1. dotnet new console -o mc -f net7.0
// 2. cd mc && dotnet run

namespace MotorcycleApp
{
    public enum MotorcycleType
    {
        Sport,
        Naked,
        Touring,
        Adventure,
        DualSport,
        Motocross,
        CrossCountry,
        Trail
    }

    public class Motorcycle
    {
        public string Model { get; private set; }
        public MotorcycleType Style { get; private set; }
        public int EngineSize { get; private set; }
        public bool OffRoad { get; private set; }

        public Motorcycle(string model, MotorcycleType style, int engineSize, bool offRoad)
        {
            Model = model;
            Style = style;
            EngineSize = engineSize;
            OffRoad = offRoad;
        }

        public bool CanJump => OffRoad;

        public static Motorcycle CreateAdventure(string model, int engineSize)
        {
            var bike = new Motorcycle(model, MotorcycleType.Adventure, engineSize, offRoad: true);
            return bike;
        }

        public override string ToString()
        {
            return $"{Model} {EngineSize}, type: {Style}, off-road: {OffRoad}, can jump: {CanJump}";
        }
    }

    class Program
    {
        static List<Motorcycle> CreateSomeBikes()
        {
            var motorcycles = new List<Motorcycle>
            {
                Motorcycle.CreateAdventure("Himalayan", 410),
                Motorcycle.CreateAdventure("Tenere", 700),
                Motorcycle.CreateAdventure("KTM", 790),
                Motorcycle.CreateAdventure("Norden", 901)
            };

            return motorcycles;
        }

        static void Main(string[] args)
        {
            Console.WriteLine("Some bikes:");
            var bikes = CreateSomeBikes();
            foreach (var b in bikes)
            {
                Console.WriteLine(b);
            }
        }
    }
}
